### PR TITLE
add shs1-test testsuite

### PR DESCRIPTION
**Pull request by mixmix**
Not working yet.

The purpose of this is to get the Javascript shs implementation passing the tests that @AljoschaMeyer wrote, as these have been used to scaffold out the Go, C, Rust implementations.

### CODE REVIEW

**Comment by dominictarr** [comment]
looks good!

**Comment by mixmix** [comment]
This is running ok by the looks but the process isn't closing, so I'm not quite sure.
It's also hard to confirm that each test has got to its expected place ... not sure how to debug, time to sleep on it and maybe see if anyone europe-side has any ideas

**Comment by mixmix** [comment]
Hey @dominictarr I'm seeing this passing for shs1-test client and server tests!
Pretty exciting. This is ready for review and merge

Please try running it locally - I want to check that the executability required for some of the tests is preserved and that everything works ok on your system

File: protocol.js [hunk 1]
```diff
@@ -116,20 +117,26 @@ module.exports = function (stateless) {
...
-      var de_nonce = state.local.app_mac.slice(0, 24)
+      var encryptNonce = state.remote.app_mac.slice(0, 24)
+      var decryptNonce = state.local.app_mac.slice(0, 24)
 
       cb(null, {
         remote: state.remote.publicKey,
         //on the server, attach any metadata gathered
         //during `authorize` call
         auth: state.auth,
+        crypto: {
```
Comment (mixmix) [inline protocol.js]: I had to add this to the callback to be able to pass the shs1-test suite.
I've not documented these.
This seems ok?

**Comment by dominictarr** [comment]
the tests run but the shs1-test:client test fails

**Comment by dominictarr** [comment]
oh, sorry, it's worse. the tests fail _sometimes_
at first they were failing then I ran them more and they started passing?

**Comment by mixmix** [comment]
I had intermittent fails when I had the timeout set too low (30). I increased that to 500, and everything passed for me ever time I ran it.

Your computer is slower, can you try increasing the timeout in both tests and seeing if that fixes it for you @dominictarr ?

**Comment by mixmix** [comment]
bump @dominictarr

**Comment by dominictarr** [comment]
I didn't do anything and it still passed.... so I don't know. Timeouts in tests are generally a bad idea, I guess if you set it high enough, and log a message "if the test failed maybe set the timeout higher" so you don't waste time debugging

**Comment by mixmix** [comment]
I didn't have any choice about having timeouts in the tests - the core js functions I'm testing require that to be given as an arg yeah?

**next steps** - should I just set the timeouts to `2e3` and then merge this?

File: test/shs1-test/client.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const toPull = require('stream-to-pull-stream')
+var sodium = require('chloride')
+const { createClient } = require('../..')
+
+const alice = sodium.crypto_sign_keypair() // client
+const appKey = Buffer.from(process.argv[2], 'hex')
+const bobPublicKey = Buffer.from(process.argv[3], 'hex') // server
+const seed = null // how do we get this fro shs1testclient ?
+
+const timeout = 1e3 // I hope this is milliseconds!
```
Comment (mixmix) [inline test/shs1-test/client.js]: change this to `2e3`

File: test/shs1-test/server.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+const toPull = require('stream-to-pull-stream')
+const { createServer } = require('../..')
+
+const appKey = Buffer.from(process.argv[2], 'hex')
+const bob = { // the keypair of the server
+  secretKey: Buffer.from(process.argv[3], 'hex'),
+  publicKey: Buffer.from(process.argv[4], 'hex')
+}
+const authorize = (pubKey, cb) => cb(null, true) // all clients are allowed to connect
+const timeout = 1e3 // I hope this is milliseconds!
```
Comment (mixmix) [inline test/shs1-test/server.js]: change this to `2e3`?

**Comment by mixmix** [comment]
The other chance is that there's a genuine intermittent problem in here which has nothing to do with timeouts ... and then there's something very wrong in either the test suite (unlikely) or shs

**Comment by dominictarr** [comment]
yeah, i'm pretty sure it's just that sometimes it takes a while to
start the node process. I suggest 10e3 instead I think. the api it
self has a time out, because that's the only way to detect if a remote
server has really died, but a local server is different. hmm, maybe -1
waits forever (memory is fuzzy...)

**Comment by mixmix** [comment]
ok, bumped up to 10e3

Can you have a look at the other changes and then merge / publish if it's fine.
Note I change some of the data coming through a callback to be able to work with the shs1-test suite

**Comment by dominictarr** [comment]
merged into 1.1.20

**Comment by mixmix** [comment]
Woop!
This was fun

**Pull request merged**